library/std/src/io/buffered/bufwriter.rs

```diff
@@ -291,64 +374,151 @@ impl<W: Write> BufWriter<W> {
-        if buf.len() >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write_all(buf);
-            self.panicked = false;
-            r
-        } else {
-            self.buf.extend_from_slice(buf);
-            Ok(())
+
+        // If, at this point, the buffer is full, we may as well eagerly
+        // attempt to flush, so that the next write will have an empty
+        // buffer.
+        if self.available() == 0 {
+            self.flush_buf()?;
         }
+
+        Ok(())
     }
 
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
-        let total_len = bufs.iter().map(|b| b.len()).sum::<usize>();
-        if self.buf.len() + total_len > self.buf.capacity() {
-            self.flush_buf()?;
-        }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
-        if total_len >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write_vectored(bufs);
-            self.panicked = false;
-            r
+        if let Some(buf) = only_one(bufs, |b| !b.is_empty()) {
+            // If there's exactly 1 incoming buffer, `Self::write` can make
+            // use of self.inner.write_vectored to attempt to combine flushing
+            // the existing buffer with writing the new one.
+            self.write(buf)
+        } else if self.get_ref().is_write_vectored() {
+            let total_len: usize = bufs.iter().map(|buf| buf.len()).sum();
+
+            if total_len > self.available() {
+                self.flush_buf()?;
+            }
+
+            if total_len >= self.capacity() {
+                self.get_mut().write_vectored(bufs)
+            } else {
+                // Correctness note: we've already verified that none of these
+                // will overflow the buffer, because total_len < capacity
+                bufs.iter().for_each(|buf| self.buf.extend_from_slice(buf));
+                Ok(total_len)
+            }
         } else {
-            bufs.iter().for_each(|b| self.buf.extend_from_slice(b));
-            Ok(total_len)
+            // Because the inner writer doesn't have native vectored write
+            // support, we should take care of buffering together the individual
+            // incoming bufs, even if the *total* length is larger than our
+            // buffer. We only want to skip our buffer if an *individual* write
+            // exceeds our buffer capacity.
+            let mut total_buffered = 0;
+
+            for buf in bufs {
+                if total_buffered == 0 {
+                    if buf.len() > self.available() {
+                        // If an individual write would overflow our remaining
+                        // capacity and we haven't buffered anything yet,
+                        // pre-flush before buffering (same as with regular
+                        // write()).
+                        self.flush_buf()?;
```

Comment:
I think this flushing in a loop violates the contract on `write_vectored`, which says it should be equivalent to `write` with the buffers concatenated, i.e. it should try to perform an underlying write at most once.

Check #78768 for an alternative implementation that does not do repetitive fill-flush cycles, at the cost of being potentially suboptimal for slices with lengths on the order of the buffer capacity, but not larger than it. I think that case would be a wrong use of `BufWriter` to worth optimizing for.

Reply:
It *does* only try to write the newly submitted data at most once. `flush_buf` only targets old data.

Reply:
Ah sorry, I missed that.